Refactor OrderDetailsViewController and OrderDetailsViewModel

ctarda opened the pull request:
Fixes #1035 

This is a pretty large PR, that requires extensive testing. My apologies for both the length and the extra testing, but there was no way I could think of to make this PR smaller.

I am opening this PR as a draft, first so that I can finally put it out for review, and second so that we have extra room to discuss the way I have refactored the existing code. I considered two alternatives:
1. Separate the view model by "verticals" (on class that fetches shipment trackings and knows how to populate the cells for those trackings, another class that does the same with products...)
2. Make the view model fetch data (products, shipment trackings, etc), pass all that data to a single class that populates all types of cells. This is what I went with in the end, I am still not 100% convinced it is the best possible approach.

## Changes
<img src="https://user-images.githubusercontent.com/2722505/60695692-3d1dbf00-9f15-11e9-9342-f70b2df20da1.jpg" width="350"/>

We would add an extra piece to the existing VC<>VM pair, that would implement `UITableViewDataSource`. 

So, after the refactoring, the responsibilities would be:
- OrderDetailsViewController just renders data and presents Notices
- OrderDetailsViewModel encapsulates data fetching and coordinates data, passing it to a data source
- OrderDetailsDataSource populates cells 

In terms of lines of code, OrderDetailsViewController goes from 1398 lines down to 468, OrderDetailsViewModel grows from 166 to 445 lines. OrderDetailsDataSource is at 772 lines.

OrderDetailsDataSource is still a bit too large, and it can be split even further, but we could do that in a follow up PR, if we decide to follow this approach.

## Testing
- Checkout the branch, navigate to the Order list
- Navigate in an out of orders, both with and without a network connection. 
- Tap all the things: fulfillment, filters, product details... we need to make sure everything works as it used to

mindgraffiti commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,445 @@
+        }
+    }
+
+    /// Helper
```
Nitpick: I think this should be plural, `Helpers`

mindgraffiti commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,445 @@
+}
+
+
+// MARK: - Syning data. Yosemite related stuff
```
unrelated nit on spelling: `Syncing` instead of `Syning`

mindgraffiti commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,445 @@
+
+// MARK: - Syning data. Yosemite related stuff
+extension OrderDetailsViewModel {
+    func syncOrder(onCompletion: ((Order?, Error?) -> ())? = nil) {
```
It's interesting that the data syncing is happening in the ViewModel. You talked about maybe adding a presenter once and now I'm starting to see why we'd need one (or something similar to it, broken out into a separate file). It's not really a ViewModel's responsibility to sync data.

ctarda replied:
Exactly. In this case, I am using the ViewModel suffix with a lot of latitude. This class is doing more than the traditional role of a view model, but I decided to stop with the refactoring here, to keep it manageable. Eventually, we'd see the need to extract the syncing, but at least I feel like now we are in a better position to do so.

mindgraffiti replied:
>I decided to stop with the refactoring here, to keep it manageable. 
yep :) I appreciate it too! This code is teaching me so many things 😁

mindgraffiti commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderEmailComposer.swift:
```diff
@@ -0,0 +1,36 @@
+import Yosemite
+import MessageUI
+
+/// Encapsulates logic to share an Order via email
```
Ahh, much better!

mindgraffiti commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderMessageComposer.swift:
```diff
@@ -0,0 +1,30 @@
+
+/// Encapsulates logic necessary to share an Order via Message
+///
+final class OrderMessageComposer: NSObject, MFMessageComposeViewControllerDelegate {
```
Such perfect, single-responsibility classes 😁

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/Cells/OrderTableViewCell.swift:
```diff
@@ -4,7 +4,7 @@ import Yosemite
 // MARK: - OrderTableViewCell
 //
-class OrderTableViewCell: UITableViewCell {
+final class OrderTableViewCell: UITableViewCell {
```
thank you!

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/Cells/SummaryTableViewCell.swift:
```diff
@@ -51,21 +56,11 @@ final class SummaryTableViewCell: UITableViewCell {
-            paymentStatusLabel.text = viewModel.order.statusKey
-        }
+    func display(presentation: SummaryTableViewCellPresentation) {
+        paymentStatusLabel.applyStyle(for: presentation.status)
```
This method removes the `else` statement that makes sure the unsupported custom status labels get styled correctly. Please restore the original if/else code.

How this cell should render unsupported, custom status labels:
<img width="354" alt="Screen Shot 2019-07-08 at 4 28 57 PM" src="https://user-images.githubusercontent.com/1062444/60843917-8ed87a80-a19d-11e9-96b8-74bf3fdfae66.png">

How this cell is currently rendering unsupported, custom status labels:
<img width="360" alt="Screen Shot 2019-07-08 at 4 29 03 PM" src="https://user-images.githubusercontent.com/1062444/60843945-a6176800-a19d-11e9-9929-86e56125d407.png">

ctarda replied:
Nice catch, sorry about that. I tried (and failed) to remove that responsibility from the cell (notice how we used to pass the view model and now we pass something of type `SummaryTableViewCellPresentation`. If you don't mind, I'd prefer to give it another try and to fix my bug, and keep that logic out of the cell. If I can't I'll roll the change back.

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/OrderDetailsViewController.swift:
```diff
@@ -191,38 +123,22 @@ private extension OrderDetailsViewController {
-
-    func configureTrackingResultsController() {
-        trackingResultsController.onDidChangeContent = { [weak self] in
+    private func prepareViewModel() {
```
Good eye. I don't think we have a standard yet. I think `configure` is our most popular use, but I'll follow whatever naming convention you and @ctarda prefer.

ctarda replied:
Yeah, `prepare` might be a Cesar-ism... sorry about that. In this case I think we'd go with configure, if only for consistency with the naming in the rest of the class.

As a side note, I am terrible with naming and naming conventions, so I'll be happy to follow whatever you folks decide.

ctarda commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,445 @@
+    }
+
+    private func configureTrackingResultsController(onReload: @escaping () -> Void) {
+        trackingResultsController.onDidChangeContent = { [weak self] in
```
I am not sure I follow @jaclync could you elaborate please?

jaclync replied:
like
```
class ResultsController<T: ResultsControllerMutableType> {
  ...
  public var onDidChangeContent: ((ResultsController) -> Void)?
}
```
so that this use case could be:
```
trackingResultsController.onDidChangeContent = { [weak self] resultsController in
  self?.dataSource.orderTracking = resultsController.fetchedObjects ?? []
  ...
}
```
just a slight prevent from calling the incorrect results controller, but pretty minor

ctarda commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource.swift:
```diff
@@ -0,0 +1,771 @@
+        }
+    }
+
+    func configureBillingAddress(cell: CustomerInfoTableViewCell) {
```
Definitely! Since all of them are already declared in an extension, I'll make the extension private

ctarda commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource.swift:
```diff
@@ -0,0 +1,771 @@
+        cell.display(presentation: presentation)
+    }
+
+    func lookUpOrderStatus(for order: Order) -> OrderStatus? {
```
Those two are consumed by the view model, so I'll extract them to a separate extension, leaving it with the default access modifier

ctarda commented:
@jaclync @mindgraffiti I think this is ready for another round of review:

- Order statuses should be rendered correctly everywhere
- Headers should be rendering correctly
- No empty details view controller
- The app should not crash when pulling down to refresh on OrderDetailsViewController

I know this is a lot to review and test, so I really appreciate your help! 🙇

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,408 @@
+//    ///
+//    private lazy var orderListener: EntityListener<Order> = {
+//        return EntityListener(storageManager: AppDelegate.shared.storageManager, readOnlyEntity: order)
+//    }()
```
do we still want this code, or this is commented out for an upcoming PR?

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,408 @@
+extension OrderDetailsViewModel {
+    func configureResultsControllers(onReload: @escaping () -> Void) {
+        dataSource.configureResultsControllers(onReload: onReload)
+        //configureOrderListener()
```
same question on commented out code :)

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,408 @@
+//            let orderStatus = self.lookUpOrderStatus(for: order)
+//            self.orderStatus = orderStatus
+//        }
+//    }
```
same on the commented out code :)

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,408 @@
+
+        UIApplication.shared.open(phoneURL, options: [:], completionHandler: nil)
+        WooAnalytics.shared.track(.orderContactAction, withProperties: ["id": order.orderID,
+                                                                        "status": self.order.statusKey,
```
nit: `self` might not be necessary?

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,408 @@
+                                                                        }
+
+                                                                        WooAnalytics.shared.track(.orderTrackingLoaded, withProperties: ["id": orderID])
+                                                                        //self?.refreshOrderTracking()
```
do we still need this code?

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsViewModel.swift:
```diff
@@ -0,0 +1,408 @@
+
+                return
+            }
+//            self?.refreshOrderProducts()
```
same question on this commented out code :)

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/Cells/OrderTableViewCell.swift:
```diff
@@ -24,11 +24,13 @@ class OrderTableViewCell: UITableViewCell {
         totalLabel.text = viewModel.totalFriendlyString
 
-        if let orderStatus = viewModel.orderStatus {
+        //let status = viewModel.orderStatus
```
do we still need this?

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/OrderLoaderViewController.swift:
```diff
@@ -182,8 +182,9 @@ private extension OrderLoaderViewController {
-        let orderStatus = lookUpOrderStatus(for: order.statusKey)
-        detailsViewController.viewModel = OrderDetailsViewModel(order: order, orderStatus: orderStatus)
+//        let orderStatus = lookUpOrderStatus(for: order.statusKey)
+//        detailsViewController.viewModel = OrderDetailsViewModel(order: order, orderStatus: orderStatus)
```
some commented out code :)

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/OrderSearchViewController.swift:
```diff
@@ -349,9 +350,9 @@ private extension OrderSearchViewController {
     func detailsViewModel(at indexPath: IndexPath) -> OrderDetailsViewModel {
         let order = resultsController.object(at: indexPath)
-        let orderStatus = lookUpOrderStatus(for: order)
+        //let orderStatus = lookUpOrderStatus(for: order)
```
seems like we don't need this?

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/OrderDetailsViewController.swift:
```diff
@@ -426,370 +234,45 @@ extension OrderDetailsViewController {
 
-            let orderStatus = self.lookUpOrderStatus(for: order)
-            self.viewModel = OrderDetailsViewModel(order: order, orderStatus: orderStatus)
+            //let orderStatus = self.viewModel.lookUpOrderStatus(for: order)
```
commented out code here :)

ctarda commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource.swift:
```diff
@@ -0,0 +1,799 @@
+        cell.taxesView.isHidden = taxesValue == nil
+
+        cell.totalLabel.text = Titles.totalLabel
+        cell.totalValue.text = totalValue
```
I agree, that is a lot of code that should probably go into the actual cell. I decided to not go that deep into the refactoring this time, to keep the scope contained, but just added an issue for it: #1116

ctarda commented:
Thanks @mindgraffiti and @jaclync , and sorry for all the cleanup I didn't do and should have done 😊 

I think I incorporated everything, with one exception: I added an issue to extract the view model mentioned to PaymentTableViewCell so that I don't break anything now. I'll attack it (#1116 ) asap.

I'll merge when the build is completed.
